src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+            prev_value = _prev_value.view(bsz * self.num_heads, -1, self.head_dim)
+            if static_kv:
+                v = prev_value
+            else:
+                assert v is not None
+                v = torch.cat([prev_value, v], dim=1)
+        prev_key_padding_mask = None  # type: Optional[Tensor]
+        if "prev_key_padding_mask" in saved_state:
+            prev_key_padding_mask = saved_state["prev_key_padding_mask"]
+        assert k is not None and v is not None
+        key_padding_mask = SelfAttention._append_prev_key_padding_mask(
+            key_padding_mask=key_padding_mask,
+            prev_key_padding_mask=prev_key_padding_mask,
+            batch_size=bsz,
+            src_len=k.size(1),
+            static_kv=static_kv,
+        )
+        saved_state["prev_key"] = k.view(bsz, self.num_heads, -1, self.head_dim)
+        saved_state["prev_value"] = v.view(bsz, self.num_heads, -1, self.head_dim)
+        saved_state["prev_key_padding_mask"] = key_padding_mask
+        # In this branch incremental_state is never None
+        assert incremental_state is not None
+        self._set_input_buffer(incremental_state, saved_state)
+        return k, key_padding_mask, v
+
+    @staticmethod
+    def _append_prev_key_padding_mask(
+        key_padding_mask: Optional[Tensor],
+        prev_key_padding_mask: Optional[Tensor],
+        batch_size: int,
+        src_len: int,
+        static_kv: bool,
+    ) -> Optional[Tensor]:
+        # saved key padding masks have shape (bsz, seq_len)
+        if prev_key_padding_mask is not None and static_kv:
+            new_key_padding_mask = prev_key_padding_mask
+        elif prev_key_padding_mask is not None and key_padding_mask is not None:
+            new_key_padding_mask = torch.cat([prev_key_padding_mask.float(), key_padding_mask.float()], dim=1)
+        # During incremental decoding, as the padding token enters and
+        # leaves the frame, there will be a time when prev or current
+        # is None
+        elif prev_key_padding_mask is not None:
+
+            filler = torch.zeros(batch_size, src_len - prev_key_padding_mask.size(1))
+            if prev_key_padding_mask.is_cuda:
+                filler = filler.cuda()
+            new_key_padding_mask = torch.cat([prev_key_padding_mask.float(), filler.float()], dim=1)
+        elif key_padding_mask is not None:
+            filler = torch.zeros(batch_size, src_len - key_padding_mask.size(1))
+            if key_padding_mask.is_cuda:
+                filler = filler.cuda()
+            new_key_padding_mask = torch.cat([filler.float(), key_padding_mask.float()], dim=1)
+        else:
+            new_key_padding_mask = prev_key_padding_mask
+        return new_key_padding_mask
+
+    def reorder_incremental_state(self, incremental_state: Dict[str, Dict[str, Optional[Tensor]]], new_order):
+        """Reorder buffered internal state (for incremental generation)."""
+        # TODO(SS): Where is this used?
```

Comment:
In our case, it will also be part of the `generate()` method of the base class `PretrainedModel` so it can very likely be removed from here.